batchflow/research/research.py

```diff
@@ -354,28 +354,60 @@ def run(self, name=None, workers=1, branches=1, n_iters=None, devices=None, exec
         self.results = ResearchResults(self.name, self.dump_results)
         self.profiler = ResearchProfiler(self.name, self.profile)
 
-        def _run():
-            self.monitor.start(self.dump_results)
+        def _start_distributor():
+            self.monitor.start(self.dump_results and self.debug)
             self.distributor.run()
             self.monitor.stop()
 
-        if detach:
-            self.process = mp.Process(target=_run)
-            self.process.start()
-            self.logger.info(f"Detach research[pid:{self.process.pid}]")
-        else:
-            _run()
+        try:
+            if detach:
+                self.process = mp.Process(target=_start_distributor)
+                self.process.start()
+                self.logger.info(f"Detach research[pid:{self.process.pid}]")
+            else:
+                _start_distributor()
+                self.terminate()
+        except KeyboardInterrupt as e:
+            self.logger.info("Research has been stopped by KeyboardInterrupt.")
+            self.terminate()
+            raise e
         return self
 
     def terminate(self):
-        """ Kill detached process. """
+        """ Kill all research processes. """
+        self.logger.info("Terminate research processes")
+
+        if self.monitor is not None:
+            self.monitor.stop(wait=False)
+
         if self.process is not None:
-            self.logger.info(f"Terminate research process[pid:{self.process.pid}]")
-            parent = psutil.Process(self.process.pid)
+            processes_to_kill = [self.process]
+        elif self.distributor is not None:
+            processes_to_kill = self.distributor.worker_processes
+        else:
+            processes_to_kill = []
+
+        for parent in processes_to_kill:
+            parent = psutil.Process(parent.pid)
             for child in parent.children(recursive=True):
+                self.logger.info(f"Terminate process[pid:{child.pid}]")
```

Comment:
Differentiate the logs: `Terminating child process: [parent.pid:], [child.pid]` and `Terminating worker's process: [parent.pid:]`

Reply:
It is difficult to distinguish between the processes:

- if `detach=True` and `parallel=True`: `parent` is detached process and and its children are workers, tasks and possibly their children (we can't destinguish them with `parent.children(recursive=True)`)
- if `detach=True` and `parallel=False`: `parent` is detached process without children
- if `detach=False` and `parallel=True`: `parent` is each of the workers and children are tasks and possibly their children
- if `detach=False` and `parallel=False`: no processes